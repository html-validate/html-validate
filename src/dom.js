'use strict';

const DOMNode = require('./domnode').default;

class DOM {
	constructor(){
		this.root = DOMNode.rootNode();
		this.active = this.root;
	}

	pushActive(node){
		this.active = node;
	}

	popActive(){
		this.active = this.active.parent;
	}

	getActive(){
		return this.active;
	}
}

module.exports = DOM;
