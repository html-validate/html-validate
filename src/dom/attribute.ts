import { Location } from "../context";
import { DynamicValue } from "./dynamic-value";

interface StaticAttribute {
	readonly value: string | null;
}

interface DynamicAttribute {
	readonly value: DynamicValue;
}

/**
 * Type guard for [[Attribute]] testing if the value is static (missing value is
 * considered static as it is a fixed known value).
 *
 * @public
 */
export function isStaticAttribute(attr: Attribute | null): attr is Attribute & StaticAttribute {
	return Boolean(attr && attr.isStatic);
}

/**
 * Type guard for [[Attribute]] testing if the value is dynamic.
 *
 * @public
 */
export function isDynamicAttribute(attr: Attribute | null): attr is Attribute & DynamicAttribute {
	return Boolean(attr && attr.isDynamic);
}

/**
 * DOM Attribute.
 *
 * Represents a HTML attribute. Can contain either a fixed static value or a
 * placeholder for dynamic values (e.g. interpolated).
 */
export class Attribute {
	/** Attribute name */
	public readonly key: string;
	public readonly value: string | DynamicValue | null;
	public readonly keyLocation: Location;
	public readonly valueLocation: Location | null;
	public readonly originalAttribute?: string;

	/**
	 * @param key - Attribute name.
	 * @param value - Attribute value. Set to `null` for boolean attributes.
	 * @param keyLocation - Source location of attribute name.
	 * @param valueLocation - Source location of attribute value.
	 * @param originalAttribute - If this attribute was dynamically added via a
	 * transformation (e.g. vuejs `:id` generating the `id` attribute) this
	 * parameter should be set to the attribute name of the source attribute (`:id`).
	 */
	public constructor(
		key: string,
		value: null | string | DynamicValue | null,
		keyLocation: Location,
		valueLocation: Location | null,
		originalAttribute?: string
	) {
		this.key = key;
		this.value = value;
		this.keyLocation = keyLocation;
		this.valueLocation = valueLocation;
		this.originalAttribute = originalAttribute;

		/* force undefined to null */
		if (typeof this.value === "undefined") {
			this.value = null;
		}
	}

	/**
	 * Flag set to true if the attribute value is static.
	 */
	public get isStatic(): boolean {
		return !this.isDynamic;
	}

	/**
	 * Flag set to true if the attribute value is dynamic.
	 */
	public get isDynamic(): boolean {
		return this.value instanceof DynamicValue;
	}

	/**
	 * Test attribute value.
	 *
	 * @param pattern - Pattern to match value against. RegExp or a string (===)
	 * @param dynamicMatches - If true `DynamicValue` will always match, if false
	 * it never matches.
	 * @returns `true` if attribute value matches pattern.
	 */
	public valueMatches(pattern: RegExp, dynamicMatches?: boolean): boolean;
	public valueMatches(pattern: string, dynamicMatches?: boolean): boolean;
	public valueMatches(pattern: RegExp | string, dynamicMatches: boolean = true): boolean {
		if (this.value === null) {
			return false;
		}

		/* dynamic values matches everything */
		if (this.value instanceof DynamicValue) {
			return dynamicMatches;
		}

		/* test value against pattern */
		if (pattern instanceof RegExp) {
			return this.value.match(pattern) !== null;
		} else {
			return this.value === pattern;
		}
	}
}
