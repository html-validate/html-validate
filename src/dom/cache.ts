/**
 * @public
 */
/* eslint-disable-next-line @typescript-eslint/no-empty-object-type -- will be augmented by other modules */
export interface DOMNodeCache {}
