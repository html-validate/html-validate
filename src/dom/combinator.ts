export enum Combinator {
	DESCENDANT,
	CHILD,
	ADJACENT_SIBLING,
	GENERAL_SIBLING,
}

export function parseCombinator(combinator: string): Combinator {
	switch (combinator){
	case undefined:
	case null:
	case "":
		return Combinator.DESCENDANT;
	case ">":
		return Combinator.CHILD;
	case "+":
		return Combinator.ADJACENT_SIBLING;
	case "~":
		return Combinator.GENERAL_SIBLING;
	default:
		throw new Error(`Unknown combinator "${combinator}"`);
	}
}
