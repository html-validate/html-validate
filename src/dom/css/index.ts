export { type CSSStyleDeclaration } from "./css-style-declaration";
export { parseCssDeclaration } from "./parse-css-declaration";
