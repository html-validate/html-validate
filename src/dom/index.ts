export { Attribute } from "./attribute";
export { HtmlElement, NodeClosed } from "./htmlelement";
export { DOMNode } from "./domnode";
export { DOMTokenList } from "./domtokenlist";
export { DOMTree } from "./domtree";
export { DynamicValue } from "./dynamic-value";
export { NodeType } from "./nodetype";
export { TextNode } from "./text";
