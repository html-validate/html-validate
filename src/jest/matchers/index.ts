export { toBeValid } from "./to-be-valid";
export { toBeInvalid } from "./to-be-invalid";
export { toHTMLValidate } from "./to-htmlvalidate";
export { toHaveError } from "./to-have-error";
export { toHaveErrors } from "./to-have-errors";
export { toMatchCodeframe } from "./to-match-codeframe";
export { toMatchInlineCodeframe } from "./to-match-inline-codeframe";
