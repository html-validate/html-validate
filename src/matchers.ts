/* eslint-disable @typescript-eslint/no-namespace, @typescript-eslint/ban-ts-comment, prefer-template, sonarjs/no-duplicate-string */

import diff from "jest-diff";
import deepmerge from "deepmerge";
import { TokenType } from "./lexer";
import { Message, Report, Result } from "./reporter";
import { ConfigData } from "./config";
import HtmlValidate from "./htmlvalidate";

interface TokenMatcher {
	type: TokenType;
	location?: any;
	data?: any;
}

declare global {
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/ban-types */
		interface Matchers<R, T = {}> {
			toBeValid(): R;
			toBeInvalid(): R;
			toBeToken(expected: TokenMatcher): R;
			toHaveError(ruleId: string, message: string, context?: any): R;
			toHaveErrors(
				errors: Array<[string, string] | Record<string, unknown>>
			): R;

			/**
			 * Validate string or HTMLElement.
			 *
			 * Test passes if result is valid.
			 *
			 * @param config - Optional HTML-Validate configuration object.
			 */
			toHTMLValidate(config?: ConfigData): R;
		}
	}
}

/**
 * Takes all messages from all files and flattens to a single array.
 */
function flattenMessages(report: Report): Message[] {
	return report.results.reduce((aggregated: Message[], result: Result) => {
		return aggregated.concat(result.messages);
	}, []);
}

function toBeValid(
	this: jest.MatcherUtils,
	report: Report
): jest.CustomMatcherResult {
	if (report.valid) {
		return {
			pass: true,
			message: /* istanbul ignore next */ () =>
				"Result should not contain error",
		};
	} else {
		const firstError = report.results[0].messages[0];
		return {
			pass: false,
			message: () =>
				`Result should be successful but had error "${firstError.message}"`,
		};
	}
}

function toBeInvalid(
	this: jest.MatcherUtils,
	report: Report
): jest.CustomMatcherResult {
	if (report.valid) {
		return {
			pass: false,
			message: () => "Result should be invalid but had no errors",
		};
	} else {
		return {
			pass: true,
			message: /* istanbul ignore next */ () =>
				"Result should not contain error",
		};
	}
}

function toHaveError(
	this: jest.MatcherUtils,
	report: Report,
	ruleId: any,
	message: any,
	context?: any
): jest.CustomMatcherResult {
	const actual = flattenMessages(report);
	const expected: any = { ruleId, message };
	if (context) {
		expected.context = context;
	}

	const matcher = [expect.objectContaining(expected)];
	const pass = this.equals(actual, matcher);
	const diffString = diff(matcher, actual, { expand: this.expand });
	const resultMessage = (): string =>
		this.utils.matcherHint(".toHaveError") +
		"\n\n" +
		"Expected token to equal:\n" +
		`  ${this.utils.printExpected(matcher)}\n` +
		"Received:\n" +
		`  ${this.utils.printReceived(actual)}` +
		/* istanbul ignore next */ (diffString
			? `\n\nDifference:\n\n${diffString}`
			: "");

	return { pass, message: resultMessage };
}

function toHaveErrors(
	this: jest.MatcherUtils,
	report: Report,
	errors: Array<[string, string] | Record<string, unknown>>
): jest.CustomMatcherResult {
	const actual = flattenMessages(report);
	const matcher = errors.map((entry) => {
		if (Array.isArray(entry)) {
			const [ruleId, message] = entry;
			return expect.objectContaining({ ruleId, message });
		} else {
			return expect.objectContaining(entry);
		}
	});
	const pass = this.equals(actual, matcher);
	const diffString = diff(matcher, actual, { expand: this.expand });
	const resultMessage = (): string =>
		this.utils.matcherHint(".toHaveErrors") +
		"\n\n" +
		"Expected token to equal:\n" +
		`  ${this.utils.printExpected(matcher)}\n` +
		"Received:\n" +
		`  ${this.utils.printReceived(actual)}` +
		/* istanbul ignore next */ (diffString
			? `\n\nDifference:\n\n${diffString}`
			: "");

	return { pass, message: resultMessage };
}

function toHTMLValidate(
	this: jest.MatcherUtils,
	// @ts-ignore DOM library not available
	actual: string | HTMLElement,
	userConfig?: ConfigData,
	filename?: string
): jest.CustomMatcherResult {
	// @ts-ignore DOM library not available
	if (actual instanceof HTMLElement) {
		actual = actual.outerHTML;
	}

	const defaultConfig = {
		rules: {
			/* jsdom normalizes style so disabling rule when using this matcher or it
			 * gets quite noisy when configured with self-closing */
			"void-style": "off",
		},
	};
	const config = deepmerge(defaultConfig, userConfig || {});

	const actualFilename = filename || this.testPath;
	const htmlvalidate = new HtmlValidate();
	const report = htmlvalidate.validateString(actual, actualFilename, config);
	const pass = report.valid;
	if (pass) {
		return { pass, message: () => "HTML is valid when an error was expected" };
	} else {
		const errors = report.results[0].messages.map(
			(message) => `  ${message.message} [${message.ruleId}]`
		);
		return {
			pass,
			message: () =>
				["Expected HTML to be valid but had the following errors:", ""]
					.concat(errors)
					.join("\n"),
		};
	}
}

function toBeToken(
	this: jest.MatcherUtils,
	actual: any,
	expected: any
): jest.CustomMatcherResult {
	const token = actual.value;

	// istanbul ignore next: TokenMatcher requires "type" property to be set, this is just a failsafe
	if (token.type) {
		token.type = TokenType[token.type];
	}

	// istanbul ignore next: TokenMatcher requires "type" property to be set, this is just a failsafe
	if (expected.type) {
		expected.type = TokenType[expected.type];
	}

	const matcher = expect.objectContaining(expected);
	const pass = this.equals(token, matcher);
	const diffString = diff(matcher, token, { expand: this.expand });
	const message = (): string =>
		this.utils.matcherHint(".toBeToken") +
		"\n\n" +
		"Expected token to equal:\n" +
		`  ${this.utils.printExpected(matcher)}\n` +
		"Received:\n" +
		`  ${this.utils.printReceived(token)}` +
		/* istanbul ignore next */ (diffString
			? `\n\nDifference:\n\n${diffString}`
			: "");

	return { pass, message };
}

expect.extend({
	toBeValid,
	toBeInvalid,
	toHaveError,
	toHaveErrors,
	toHTMLValidate,
	toBeToken,
});
