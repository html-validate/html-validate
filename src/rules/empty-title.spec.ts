import { DynamicValue, HtmlElement } from "../dom";
import HtmlValidate from "../htmlvalidate";
import "../matchers";

describe("rule empty-title", () => {
	let htmlvalidate: HtmlValidate;

	beforeAll(() => {
		htmlvalidate = new HtmlValidate({
			rules: { "empty-title": "error" },
		});
	});

	it("should not report when title has text", () => {
		expect.assertions(1);
		const report = htmlvalidate.validateString("<title>lorem ipsum</title>");
		expect(report).toBeValid();
	});

	it("should not report when title has children with text", () => {
		expect.assertions(1);
		const report = htmlvalidate.validateString(
			"<title><span>lorem ipsum</span></title>"
		);
		expect(report).toBeValid();
	});

	it("should not report when title has dynamic text", () => {
		expect.assertions(1);
		function processElement(node: HtmlElement): void {
			node.appendText(new DynamicValue(""));
		}
		const report = htmlvalidate.validateString("<title></title>", null, {
			processElement,
		});
		expect(report).toBeValid();
	});

	it("should report error when title has no text content", () => {
		expect.assertions(2);
		const report = htmlvalidate.validateString("<title></title>");
		expect(report).toBeInvalid();
		expect(report).toHaveError(
			"empty-title",
			"<title> cannot be empty, must have text content"
		);
	});

	it("should report error when title has no children with content", () => {
		expect.assertions(2);
		const report = htmlvalidate.validateString("<title><span></span></title>");
		expect(report).toBeInvalid();
		expect(report).toHaveError(
			"empty-title",
			"<title> cannot be empty, must have text content"
		);
	});

	it("should report error when title only has whitespace content", () => {
		expect.assertions(2);
		const report = htmlvalidate.validateString("<title> </title>");
		expect(report).toBeInvalid();
		expect(report).toHaveError(
			"empty-title",
			"<title> cannot be empty, must have text content"
		);
	});

	it("should report error when title only has comment", () => {
		expect.assertions(2);
		const report = htmlvalidate.validateString(
			"<title>\n<!-- foo -->\n</title>"
		);
		expect(report).toBeInvalid();
		expect(report).toHaveError(
			"empty-title",
			"<title> cannot be empty, must have text content"
		);
	});

	it("smoketest", () => {
		expect.assertions(1);
		const report = htmlvalidate.validateFile(
			"test-files/rules/empty-title.html"
		);
		expect(report.results).toMatchSnapshot();
	});

	it("should contain documentation", () => {
		expect.assertions(1);
		expect(htmlvalidate.getRuleDocumentation("empty-title")).toMatchSnapshot();
	});
});
