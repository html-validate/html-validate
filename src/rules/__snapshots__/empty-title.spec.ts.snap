// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule empty-title should contain documentation 1`] = `
Object {
  "description": "The \`<title>\` element cannot be empty, it must have textual content.

It is used to describe the document and is shown in the browser tab and titlebar.
WCAG and SEO requires a descriptive title and preferably unique within the site.

Whitespace is ignored.",
  "url": "https://html-validate.org/rules/empty-title.html",
}
`;

exports[`rule empty-title smoketest 1`] = `
Array [
  Object {
    "errorCount": 2,
    "filePath": "test-files/rules/empty-title.html",
    "messages": Array [
      Object {
        "column": 2,
        "context": undefined,
        "line": 2,
        "message": "<title> cannot be empty, must have text content",
        "offset": 28,
        "ruleId": "empty-title",
        "ruleUrl": "https://html-validate.org/rules/empty-title.html",
        "selector": "title:nth-child(2)",
        "severity": 2,
        "size": 5,
      },
      Object {
        "column": 2,
        "context": undefined,
        "line": 3,
        "message": "<title> cannot be empty, must have text content",
        "offset": 44,
        "ruleId": "empty-title",
        "ruleUrl": "https://html-validate.org/rules/empty-title.html",
        "selector": "title:nth-child(3)",
        "severity": 2,
        "size": 5,
      },
    ],
    "source": "<title>Lorem ipsum</title>
<title></title>
<title> </title>
",
    "warningCount": 0,
  },
]
`;
